src/allmydata/test/web/test_root.py

```diff
@@ -36,3 +45,363 @@ def test_missing(self):
         self.assertIdentical(res, ctx)
         self.assertEqual(ctx.slots["version"], "")
         self.assertEqual(ctx.slots["nickname"], "")
+
+
+class FakeUploader(object):
+    """
+    """
+    def get_helper_info(self):
+        return ("furl", False)
+
+
+class FakeHelper(object):
+    def get_stats(self):
+        return {
+            "chk_upload_helper.active_uploads": 0,
+        }
+
+
+class FakeStorageBroker(object):
+    def get_connected_servers(self):
+        return {}
+
+    def get_all_serverids(self):
+        return ()
+
+    def get_known_servers(self):
+        return ()
+
+
+class FakeMagicFolder(object):
+    def get_public_status(self):
+        return (True, ["this magic folder is alive"])
+
+
+# XXX there are several 'fake client' instance throughout the code
+# .. probably should be a single one that works better and covers all
+# the cases etc.
+class FakeClient(object):
+    """
+    just enough to let the node acquire a downloader (which it won't
+    use), and to get default encoding parameters
+    """
+    nickname = "fake_nickname"
+    stats_provider = None
+    uploader = FakeUploader()
+    helper = FakeHelper()
+    storage_broker = FakeStorageBroker()
+
+    _secret_holder = SecretHolder("lease secret", "convergence secret")
+    _magic_folders = {
+        "foo": FakeMagicFolder(),
+    }
+
+    def get_long_nodeid(self):
+        return "v0-nodeid"
+
+    def get_long_tubid(self):
+        return "v0-tubid"
+
+    def introducer_connection_statuses(self):
+        return {}
+
+    def get_auth_token(self):
+        return "x"
+
+    def getServiceNamed(self, name):
+        return {
+            "uploader": self.uploader,
+            "helper": self.helper,
+        }[name]
+
+    def get_encoding_parameters(self):
+        return {"k": 3, "n": 10}
+
+    def get_storage_broker(self):
+        return self.storage_broker
+
+    def get_history(self):
+        return None
+
+
+
+class RenderRoot(unittest.TestCase):
+
```

Comment:
a docstring here would be nice, and on all of the test methods

Reply:
deleted